README.md
# kamihubuki-js

kamihubuki-js is a lightweight 2D confetti physics engine in Javascript and TypeScript.

![confetti](./confetti-anim.gif)

## Install

You can install using package manager.

```shell
npm install blackspotbear/kamihubuki-js
```

## Usage

The following is a minimal example around createing and updating confetti.

```javascript
var kh = require('kamihubuki-js');

// create a confetti instance.
var co = new kh.Confetti({
    // initial position, velocity, etc...
    x: x, y: y,
    vx: vx, vy: vy,
    angle: angle,
    fins: [ // add fins that rotate confetti.
        {
            angle: (Math.random() * 2 - 1) * (Math.PI / 2),
            size: 1,
            armAngle: 0,
            armLength: 1
        },
        {
            angle: (Math.random() * 2 - 1) * (Math.PI / 2),
            size: 2,
            armAngle: Math.random() * Math.PI,
            armLength: 1
        }
    ],
    av: 0,
    M: 0.5,
    K: 0.4,
    I: 5,
    RD: 0.99
});

// Update confetti every frame.
co.update(1 / 30, 0, 0, 0, -100);
```

More details, see [this documentation](./how-it-works.md).

## API Reference

To get API reference run:

```sh
npm run doc
```

## LICENSE

This software is released under the MIT License, see LICENSE.txt.


CHANGELOG.md
# CHANGELOG.md

## Unreleased changes

Update dependencies.

## 0.1 0

Initial commit.


src/index.ts
/**
 * Fin.
 */
export interface Fin {
	/** fin's angle relative to arm */
	angle: number;

	/** fin's size */
	size: number;

	/** arm angle relative to particle (confetti) */
	armAngle: number;

	/** arm's length */
	armLength: number;
}

/**
 * Confetti constructor parameter object.
 */
export interface ConfettiParameterObject {
	/** x position */
	x: number;

	/** y position */
	y: number;

	/** x velocity */
	vx: number;

	/** y velocity */
	vy: number;

	/** angular velocity */
	av: number;

	/** angle */
	angle: number;

	/** mass */
	M: number;

	/** air resistance coefficient */
	K: number;

	/** moment */
	I: number;

	/** angular velocity's damper [0, 1]  */
	RD: number;

	/** fins */
	fins: Fin[];
}

/**
 * Confetti.
 *
 * Y-axis goes bottom to top.
 */
export class Confetti {
	/** x position */
	x: number;

	/** y position */
	y: number;

	/** x velocity */
	vx: number;

	/** y velocity */
	vy: number;

	/** angular velocity */
	av: number;

	/** angle */
	angle: number;

	/** mass */
	M: number;

	/** air resistance coefficient */
	K: number;

	/** moment */
	I: number;

	/** angular velocity's damper [0, 1]  */
	RD: number;

	/** fins */
	fins: Fin[];

	constructor(param: ConfettiParameterObject) {
		this.x = param.x;
		this.y = param.y;
		this.vx = param.vx;
		this.vy = param.vy;
		this.av = param.av;
		this.angle = param.angle;
		this.M = param.M;
		this.K = param.K;
		this.I = param.I;
		this.RD = param.RD;
		this.fins = param.fins;
	}

	/**
	 * upadte confetti.
	 *
	 * @param dt elapse time in sec.
	 * @param fx external force x component.
	 * @param fy external formce y component.
	 * @param gx gravity x component.
	 * @param gy gravity y component.
	 */
	update(dt: number, fx: number, fy: number, gx: number, gy: number): void {
		const ff = this.calcFinForce(dt, this.fins);

		const ax = (fx + ff.fx - this.vx * this.K) / this.M + gx;
		const ay = (fy + ff.fy - this.vy * this.K) / this.M + gy;
		this.vx += ax * dt;
		this.vy += ay * dt;
		this.x += this.vx * dt;
		this.y += this.vy * dt;

		const alpha = ff.tau / this.I;
		this.av += alpha * dt;
		this.av *= this.RD;
		this.angle += this.av * dt;
	}

	private calcFinForce(dt: number, fins: Fin[]): { tau: number, fx: number, fy: number } {
		let totalTau = 0;
		let totalFx = 0;
		let totalFy = 0;

		for (let i = 0; i < fins.length; i++) {
			const fin = fins[i];

			const armWAngle = this.angle + fin.armAngle;
			const armDirX = Math.cos(armWAngle);
			const armDirY = Math.sin(armWAngle);

			const finWAngle = this.angle + fin.armAngle + fin.angle;
			const finNx = Math.cos(finWAngle);
			const finNy = Math.sin(finWAngle);

			// tip speed
			const tvx = this.vx - armDirY * this.av * fin.armLength;
			const tvy = this.vy + armDirX * this.av * fin.armLength;

			// air resistance
			const arfx = -tvx * this.K * fin.size;
			const arfy = -tvy * this.K * fin.size;

			const f = finNx * arfx + finNy * arfy;
			const fx = f * finNx;
			const fy = f * finNy;

			// torque
			totalTau += (armDirX * fy - armDirY * fx) * fin.armLength;

			const af = armDirX * fx + armDirY * fy;
			const afx = af * armDirX;
			const afy = af * armDirY;

			totalFx += afx;
			totalFy += afy;
		}

		return { tau: totalTau, fx: totalFx, fy: totalFy };
	}
}


sample.p5/src/sketch.js
const p5 = require("p5");
const kh = require("kamihubuki-js");

const sketch = p => {
	const canvasWidth = 640;
	const canvasHeight = 480;
	let confetties = [];

	const createConfetti = (x, y) => {
		const vx = 0;
		const vy = 200;
		const angle = Math.PI * p.random(Math.PI);
		const fins = [
			{
				angle: p.random(-Math.PI / 2, Math.PI / 2),
				size: 30,
				armAngle: 0,
				armLength: 1
			},
			{
				angle: p.random(-Math.PI / 2, Math.PI / 2),
				size: 30,
				armAngle: p.random(Math.PI / 2, Math.PI),
				armLength: 1
			}
		];

		const co = new kh.Confetti({
			x: x, y: y,
			vx: vx, vy: vy,
			angle: angle,
			fins: fins,
			av: 0,
			M: 0.5,
			K: 0.02,
			I: 3,
			RD: 0.99
		});

		return {
			co: co,
			color: p.color(p.random(255), p.random(255), p.random(255))
		};
	};

	p.setup = () => {
		p.frameRate(30);
		p.resizeCanvas(canvasWidth, canvasHeight);
	};

	p.draw = () => {
		for (let i = 0; i < 2; i++) {
			confetties.push(createConfetti(p.random(0, canvasWidth), p.random(-64, 0)));
		}

		confetties = confetties.filter(confetti => {
			const co = confetti.co;
			co.update(1 / p.getFrameRate(), 0, 0, 0, 200);
			return co.x >= 0 && co.x <= canvasWidth && co.y <= canvasHeight;
		});

		p.background(0xBB, 0xBB, 0xFF);

		confetties.forEach(confetti => {
			const co = confetti.co;
			p.push();
			p.translate(co.x, co.y);
			p.rotate(co.angle);
			p.fill(confetti.color);
			p.rect(-10, -5, 20, 10);
			p.pop();
		});
	};
};

new p5(sketch);


sample.ae/src/main.ts
import * as kh from "kamihubuki-js";

let windX = 0;
let windY = 0;
const colors = [
	"blue",    "navy",  "teal",   "green",
	"lime",    "aqua",  "yellow", "red",
	"fuchsia", "olive", "purple", "maroon"
];
let colorIndex = 0;

function createConfetti(x: number, y: number): kh.Confetti {
	const vx = 0;
	const vy = 200;
	const angle = g.game.random.get(0, 100) / 100 * Math.PI;

	const fins: kh.Fin[] = [
		{
			angle: g.game.random.get(-100, 100) / 100 * (Math.PI / 2),
			size: 30,
			armAngle: 0,
			armLength: 1
		},
		{
			angle: g.game.random.get(-100, 100) / 100 * (Math.PI / 2),
			size: 30,
			armAngle: g.game.random.get(50, 100) / 100 * Math.PI,
			armLength: 1
		}
	];

	const co = new kh.Confetti({
		x: x, y: y,
		vx: vx, vy: vy,
		angle: angle,
		fins: fins,
		av: 0,
		M: 0.5,
		K: 0.02,
		I: 3,
		RD: 0.99
	});

	return co;
}

function createConfettiEntity(scene: g.Scene, co: kh.Confetti, noCollision: boolean): g.E {
	const w = 20;
	const h = 10;
	const w2 = w / 2;
	const h2 = h / 2;

	const rect = new g.FilledRect({
		scene: scene,
		width: w,
		height: h,
		x: co.x - w2,
		y: co.y - h2,
		cssColor: colors[colorIndex]
	});

	colorIndex = (colorIndex + 1) % colors.length;

	rect.update.add(() => {
		co.update(1 / g.game.fps, windX, windY, 0, 200);

		// collision detection and resolution.
		if (noCollision) {
			if (co.x < 0 || co.x > g.game.width || co.y > g.game.height) {
				rect.destroy();
			}
		} else {
			const k = 0.5;
			if (co.x < 0) {
				co.x = 0;
				co.vx = Math.abs(co.vx) * k;
			} else if (co.x > g.game.width) {
				co.x = g.game.width;
				co.vx = -Math.abs(co.vx) * k;
			}
			if (co.y < -g.game.height) {
				co.y = -g.game.height;
				co.vy = 0;
			} else if (co.y > g.game.height) {
				co.y = g.game.height;
				co.vy = -Math.abs(co.vy) * k;
			}
		}

		rect.x = co.x - w2;
		rect.y = co.y - h2;
		rect.angle = -co.angle / Math.PI * 180;
		rect.modified();
	});

	return rect;
}

function main(param: g.GameMainParameterObject): void {
	const scene = new g.Scene({game: g.game});

	scene.loaded.add(() => {

		const floor = new g.FilledRect({
			scene: scene,
			width: g.game.width,
			height: 4,
			x: 0,
			y: g.game.height - 4,
			cssColor: "green"
		});

		let pointDownCaptureCntr = 0;
		let pointDown = false;
		let mouseX = 0;
		let mouseY = 0;
		scene.pointDownCapture.add(ev => {
			if (ev.target) {
				return;
			}
			pointDown = true;
			mouseX = ev.point.x;
			mouseY = ev.point.y;
		});
		scene.pointUpCapture.add(ev => pointDown = false);
		scene.pointMoveCapture.add(ev => {
			mouseX = ev.point.x + ev.startDelta.x;
			mouseY = ev.point.y + ev.startDelta.y;
		});

		scene.update.add(() => {
			pointDownCaptureCntr++;

			if (! pointDown) {
				return;
			}

			if (pointDownCaptureCntr % 3 !== 0) {
				return;
			}

			const co = createConfetti(mouseX, mouseY);
			const confettiEntity = createConfettiEntity(scene, co, false);
			scene.append(confettiEntity);
		});

		const fanButton = new g.FilledRect({
			scene: scene,
			x: 8,
			y: g.game.height - 8 - 32,
			width: 32,
			height: 32,
			cssColor: "blue",
			touchable: true
		});

		const fanAngle = Math.PI / 2 * 1;
		const fanForce = -2000;
		let fanButtonPressed = false;
		fanButton.pointDown.add(() => {
			fanButtonPressed = true;
			windX = fanForce * Math.cos(fanAngle);
			windY = fanForce * Math.sin(fanAngle);
		});
		fanButton.pointUp.add(() => {
			fanButtonPressed = false;
			windX = 0;
			windY = 0;
		});

		fanButton.update.add(() => {
			if (fanButtonPressed) {
				fanButton.angle += -30;
				fanButton.modified();
			}
		});

		scene.append(floor);
		scene.append(fanButton);
	});

	g.game.pushScene(scene);
}

export = main;
